### ISSUE TITLE
JobQueue observer

### ISSUE THREAD

**Issue by adam-fowler** [hummingbird-project/swift-jobs#45]
Add array of observers to queue. When events like a job is added, finished, failed occur these observers are informed. This allows us to separate out non-core job handling code into individual observers. eg 

- publishing events to service eg mqtt, rabbitmq, AWS SQS
- schedule new jobs based on jobs finishing

If we define an observer protocol there are two ways we can define this

1) A protocol that includes methods for every event
2) A protocol that has one function `onEvent` which is provided with a event type (with additional metadata)

2 has the advantages of being more flexible than 1 but it might have performance issues as every observer would have to get the event type before working out what to do with it.

@Joannis @thoven87

**Comment by thoven87** [comment]
> Add array of observers to queue. When events like a job is added, finished, failed occur these observers are informed. This allows us to separate out non-core job handling code into individual observers. eg
> 
> * publishing events to service eg mqtt, rabbitmq, AWS SQS
> * schedule new jobs based on jobs finishing
> 
> If we define an observer protocol there are two ways we can define this
> 
> 1. A protocol that includes methods for every event
> 2. A protocol that has one function `onEvent` which is provided with a event type (with additional metadata)
> 
> 2 has the advantages of being more flexible than 1 but it might have performance issues as every observer would have to get the event type before working out what to do with it.
> 
> @Joannis @thoven87

I agree with defining a protocol which includes a method for every event is more performant. What's the actual cost of the send options? Since every job would already have all possible observers by default?

**Comment by adam-fowler** [comment]
See #54 and https://github.com/hummingbird-project/hummingbird-evolution/pull/4

**Comment by adam-fowler** [comment]
Closing this as #54 was merged
